src/problems/AllPaths/algo/matrix_bfs/intersection.py

```diff
@@ -0,0 +1,189 @@
+        Create a block diagonal matrices from graph and regex matrices for each symbol
+        """
+        num_vert_graph = self.graph.get_number_of_vertices()
+        num_vert_regex = self.regular_automaton.num_states
+        diag_num_verts = num_vert_graph + num_vert_regex
+
+        diag_matrices = dict()
+        for symbol in self.regular_automaton.matrices:
+            if symbol in self.graph:
+                diag_matrix = Matrix.sparse(BOOL, diag_num_verts, diag_num_verts)
+                diag_matrix.assign_matrix(
+                    self.regular_automaton.matrices[symbol],
+                    slice(0, num_vert_regex - 1),
+                    slice(0, num_vert_regex - 1),
+                )
+                diag_matrix.assign_matrix(
+                    self.graph[symbol],
+                    slice(num_vert_regex, diag_num_verts - 1),
+                    slice(num_vert_regex, diag_num_verts - 1),
+                )
+
+                diag_matrices[symbol] = diag_matrix
+
+        return diag_matrices
+
+    def intersect_bfs(self) -> EpsilonNFA:
+        """
+        Intersection implementation with synchronous breadth first traversal
+        of a graph and regular grammar represented in automata
+        """
+        num_vert_graph = self.graph.get_number_of_vertices()
+        num_vert_regex = self.regular_automaton.num_states
+
+        num_verts_inter = num_vert_graph * num_vert_regex
+        num_verts_diag = num_vert_graph + num_vert_regex
+
+        graph = self.graph
+        regex = self.regular_automaton.matrices
+
+        regex_start_states = self.regular_automaton.start_states
+
+        diag_matrices = self.create_diag_matrices()
+
+        found = Vector.sparse(BOOL, num_verts_diag)
+        vect = Vector.sparse(BOOL, num_verts_diag)
+
+        # vectors of found nodes
+        # should be done with mask probably, not a queue
+        curr_front = list()
+        visited = Vector.sparse(BOOL, num_verts_inter)
+
+        for regex_start_st in regex_start_states:
+            for gr_start_st in range(num_vert_regex, num_verts_diag):
+                vect[regex_start_st] = True
+                vect[gr_start_st] = True
+
+                curr_front.append(vect)
+
+                while len(curr_front) > 0:
+                    vect = curr_front.pop(0)
+
+                    for symbol in regex:
+                        if symbol in graph:
+                            with semiring.LOR_LAND_BOOL:
+                                found = vect.vxm(diag_matrices[symbol])
+
+                            reg_vert, grph_vert = tuple(vect.I)
+                            prev_vert_index = self.to_inter_coord(
+                                grph_vert - num_vert_regex, reg_vert
+                            )
+                            next_vert_index = prev_vert_index
+
+                            visited[prev_vert_index] = True
+
+                            if not found.iseq(vect) and len(list(found.I)) == 2:
```

Comment:
Looks too complicated. Can it be expressed in terms of GraphBLAS vector operations?

Reply:
What if we have multiple ways to do the next step? Why the result is limited by 2?